restructure tutorial, add a learning section (#89)
FAIRmat-NFDI/nyaml: docs/learn/yaml-notation.md

# nyaml notation reference

This page is the detailed reference for writing NeXus definitions in YAML using `nyaml`. It documents all building blocks and special keywords, with YAML and NXDL XML comparisons for each.

> For a step-by-step introduction that teaches these concepts progressively through an example, see [Tutorials > Writing a NeXus definition in YAML](../tutorials/writing-a-yaml-definition.md).

Reviewer: There was somewhere else a writing "step by step", doing a `grep` for consistency is appreciated

Author: Yes, in the linked tutorial, so I think that's fine.

!!! note
    The NeXus `choice` concept is not yet supported in `nyaml`.

---

## Root section

The root section is the top-level block of a definition. It maps to the XML `<definition>` element and contains the following keys:

- `category`: `application` for application definitions, `base` for base classes
- `type`: always `group`
- `doc`: free-text description of the definition
- `symbols`: named dimension constants used throughout the definition (optional)

> - The schema name (e.g. `NXmpes(NXobject)`): the parenthesised base class indicates what this definition extends. An application definition may extend `NXobject` or another application definition.

Reviewer: ```suggestion
- The schema name (e.g. `NXmpes(NXobject)`): the parenthesized base class indicates what this definition extends. An application definition extends either `NXobject` or another application definition.
```

                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                category="application"
                type="group"
                name="NXmpes"
                extends="NXobject"
                xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd">
      <doc>
        This is the most general application definition for multidimensional photoelectron spectroscopy.
        .. _ISO 18115-1:2023: https://www.iso.org/standard/74811.html
        .. _IUPAC Recommendations 2020: https://doi.org/10.1515/pac-2019-0404
      </doc>
      <symbols>
        <doc>
          The symbols used in the schema to specify e.g. dimensions of arrays
        </doc>
        <symbol name="n_transmission_function">
          <doc>
            Number of data points in the transmission function.
          </doc>
        </symbol>
      </symbols>
    </definition>
    ```

---

## Groups

[NeXus groups](https://manual.nexusformat.org/design.html#design-groups){:target="_blank" rel="noopener"} are instances of NeXus base classes and form the compositional structure of a definition. A group is written as `name(NXbaseclass):` in YAML.

> The `nameType` keyword (see below) controls whether the name is matched exactly, loosely, or as a prefix pattern in HDF5 files. Dynamic initialisation (using `nameType: any`) allows multiple instances of the same group type at the same hierarchy level; for example, `(NXmanipulator)` (with `nameType: any`) can be instantiated as `manipulator1` and `manipulator2` during data writing.

Reviewer: ```suggestion
The `nameType` keyword (see below) controls whether the name is matched exactly, loosely, or as a prefix (or suffix) pattern in HDF5 files. Dynamic initialization (using `nameType: any`) allows multiple instances of the same group type at the same hierarchy level; for example, `(NXmanipulator)` (with `nameType: any`) can be instantiated as `manipulator1` and `manipulator2` during data writing.
```

| Mixed case           | &#10003;                   | &#10003; (with warning)        | &#10003;                     | `specified`    |
| Anonymous group name | &#10003; (with error)      | &#10003;                       | &#10003; (with error)        | `any`          |

=== "YAML"
    ```yaml
    source_TYPE(NXsource):
      exists: recommended
      nameType: partial
      doc: |
        A source used to generate a beam.
    (NXmanipulator):
      exists: optional
      nameType: any # default for anonymous groups
      doc: |
        Manipulator for positioning of the sample.
      value_log(NXlog):
        exists: optional
    ```

=== "XML"
    ```xml
    <group name="source_TYPE" type="NXsource" recommended="true" nameType="partial">
      <doc>
        A source used to generate a beam.
      </doc>
    </group>
    <!-- If no nameType is specified, NXmanipulator has nameType any by default -->
    <group type="NXmanipulator" optional="true" nameType="any">
      <doc>
        Manipulator for positioning of the sample.
      </doc>
      <group name="value_log" type="NXlog" optional="true"/>
    </group>
    ```

---

### `exists`

> Controls whether a concept must be present in a conforming file. By default, all concepts in a base class are optional and all concepts in an application definition are required.

Reviewer 2: ```suggestion
Controls whether an instance of a concept is present in a conforming NeXus file. By default, all concepts in a base class are optional, and all concepts in an application definition are required.
```

      <doc>
        This calibration procedure is used to account for the different transmission efficiencies.
      </doc>
      <group name="calibrationDATA" type="NXdata" minOccurrences="3" maxOccurrences="unbounded"/>
    </group>
    ```

In the example above, `transmission_correction` is optional, while `calibrationDATA` must appear at least three times.

---

### `unit`

Specifies the NeXus unit category for a field. Use one of the existing [unit categories](https://manual.nexusformat.org/nxdl-types.html#unit-categories-allowed-in-nxdl-specifications){:target="_blank" rel="noopener"} (e.g. `NX_VOLTAGE`, `NX_LENGTH`, `NX_WAVELENGTH`) rather than raw unit strings like `"m"`. The unit category declares the physical dimension; the actual unit written to the HDF5 file (e.g. `V`, `mV`) is stored as a sibling attribute `fieldname/@units`.

=== "YAML"
    ```yaml
    detector_voltage(NX_FLOAT):
      unit: NX_VOLTAGE
      doc: |
        Voltage applied to detector.
    ```
=== "XML"
    ```xml
    <field name="detector_voltage" type="NX_FLOAT" units="NX_VOLTAGE">
      <doc>Voltage applied to detector.</doc>
    </field>
    ```

---

### `dimensions`

> Describes the shape of a multidimensional data array. The `rank` key gives the number of dimensions. Each dimension is specified with `dim` using either the full array-of-arrays form or the shorter tuple form. Use symbolic names from the root `symbols` block rather than hardcoded integers.

Reviewer: Using the "dimensions" keyword constraints the instance data of a field to an array with at least one dimension.

Author: I added that information as well.
